### ISSUE TITLE
Destructor not called for ref member of object in specific circumstances with ARC

### ISSUE THREAD

**Issue by juliusikkala** [nim-lang/Nim#15629]
The destructor is not being called for 'inner' in the example code.

### Example
```nim
type inner = object
type outer = ref inner

proc `=destroy`(b: var inner) =
  echo "destroy"

proc newOuter(): outer =
  echo "create"
  new(result)

type holder = object
  contents: outer

proc main() = 
  # Something about this specific case causes the destructor to not be called.
  var t: holder
  t.contents = newOuter()
  # Putting the above two lines outside of the proc also causes the destructor
  # to be called correctly.
  # new(t.contents) # This works correctly
  # var u = holder(contents: newOuter()) # So does this

main()
```

### Current Output
```
create
```

### Expected Output
```
create
destroy
```

### Additional Information

```
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2020-10-18
Copyright (c) 2006-2020 by Andreas Rumpf

git hash: 78cbb52fa01d6c73307fe276bdf7d4c34d7a556c
active boot switches: -d:release
```

This problem occurs at least in 1.2.6 and 1.4.0 too, with both the C and C++ backends. Both ARC and ORC are affected, the other GC strategies don't seem to reliably call destructors anyway(?).

**Comment by cooldome** [comment]
Compiling with `--cursorInference:off` fixes the issue

**Comment by juliusikkala** [comment]
I tried `--cursorInference:off` in the actual code I was working on and was unfortunately able to derive another case from it where the destructor of "inner" does not get called even with that option. This is the exact same setup, but with a destructor for "holder" (I also took the proc main out since this one seems to break without that too!):
```nim
type inner = object
type outer = ref inner

proc `=destroy`(b: var inner) =
  echo "destroy"

proc newOuter(): outer =
  echo "create"
  new(result)

type holder = object
  contents: outer

proc `=destroy`(p: var holder) = discard

var t: holder = holder(contents: newOuter())
```
Or have I misunderstood destructors, should they call `=destroy` manually for all members? Doing that makes this example work as expected.
